### PR TITLE
Add per-model scoped killswitch threshold

### PR DESCRIPTION
**Pull request by iceteaSA**
## What

Adds a third, model-scoped dimension to the per-account killswitch: a `scoped` threshold that hard-limits traffic to a weekly scoped quota carve-out (the `weekly_scoped` "Fable only" window) **without killing the account for other models**.

When the killswitch is enabled and a request's model matches a scoped window at/below the `scoped` threshold, the request reroutes to a scoped-eligible fallback and hard-blocks with a 429 if none qualify — while Sonnet/Opus traffic on the same account continues normally. Additive to the existing `five_hour`/`seven_day` checks; generic over any scoped window (not Fable-hardcoded).

## Changes

- **`killswitchPassesPolicy(quota, storage, accountId?, modelId?)`** — new optional `modelId`. When present and a scoped window matches it, the window's `remainingPercent` is checked against the per-account `scoped` threshold. When absent, behavior is unchanged. The scoped check runs **before** the unknown-5h/7d fail-open return, so an exhausted scoped window blocks independently of missing 5h/7d windows.
- **Comparison is `<=` (inclusive)** for scoped — deliberately distinct from the strict `<` on 5h/7d — so the default threshold `0` fires at true exhaustion (matching `quotaSnapshotModelScopeIsExhausted`'s `<= 0`).
- **Config** — `KillswitchThresholds` gains an optional `scoped` key (default `0`); `normalizeKillswitchThresholds` resolves it with the same finite-guard/default pattern. Backward-compatible: existing two-value configs read `scoped` as the default.
- **Command** — `/claude-killswitch set main:80,10,0` accepts an optional third number (fh, sd, scoped). Two-number form parses identically to before. Status table gains a Scoped column.
- **Enforcement** — `modelId` is threaded into the three routing-path call sites (main killswitch gate + both fallback filters), leaving the sidebar/status callers model-free so account-level killed state is unchanged.
- **Block message** — a scoped-driven terminal 429 names the model (`"<Model> weekly limit reached, …"`) and `killswitchRetryAfterSeconds` honors the scoped window's weekly `resetsAt`. A 5h/7d-driven block keeps the account-level message (classified by the same `<= scoped threshold` check that drives the block, so message and reason never disagree).

Opencode-only, consistent with the existing killswitch (Pi has no killswitch enforcement).

## Verification

- `bun run typecheck` — clean (all packages)
- `bun run lint` — clean
- `bun test`: core 34/0, opencode 770/0, pi 47/0
- New tests cover: model-scoped isolation (same quota, non-matching model → account stays live), the `<=` boundary at the `0` default and a raised threshold, `modelId`-absent regression lock, the unknown-window × fail-open × scoped-exhausted interaction, block-message attribution (5h/7d block with a healthy scoped window → generic message; exhausted scoped window → model-named), and config round-trip for the two- and three-value forms.

---
<a href="https://app.blacksmith.sh/cortexkit/codesmith/anthropic-auth/pr/109"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1785836709&installation_id=135454708&pr_number=109&repository=cortexkit%2Fanthropic-auth&return_to=https%3A%2F%2Fgithub.com%2Fcortexkit%2Fanthropic-auth%2Fpull%2F109&signature=bc4b17c81390edf45a8cf0544060ec41c769517994d45582667475e714008a8c"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

---
## Summary by cubic
Adds a per-model `scoped` killswitch threshold to enforce weekly model carve-outs (e.g., Fable) without killing the account for other models. Routing, CLI, retry hints, and 429 messages are scope-aware with correct attribution and weekly reset timing; tests lock the priority rules.

- **New Features**
  - Add `scoped` threshold (default 0); inclusive `<=`; evaluated when a request includes `modelId`; additive to 5h/7d.
  - Thread `modelId` through the main gate and fallback filters; exhausted scoped windows reroute to scoped-eligible fallbacks or 429.
  - Update retry hints to accept a `scopedModelId` and use only the matched weekly reset; extend `/claude-killswitch set` to accept a third number (fh, sd, scoped); status table shows a Scoped column; thresholds normalized; block messages name the model for scoped-driven blocks.

- **Bug Fixes**
  - Run the scoped check before the unknown 5h/7d fail-open decision (MUST-1).
  - Correct 429 attribution: only label a block as model-scoped when the matched scoped window is at/below the `scoped` threshold and 5h/7d did not already kill the account (MUST-2, FINDING 2).
  - Fix retry-after for scoped blocks: when `scopedModelId` is provided, consider only the matched scoped window’s `resetsAt` so the hint reflects the weekly reset (FINDING 1).
  - Strengthen the MUST-2 test to actually trigger account-level blocks by setting 5h/7d remaining below the strict `<` thresholds (ensures the priority guard is exercised).

<sup>Written for commit 660e47dbc49699c957ed82ca36d3b8be8a9d380e. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/cortexkit/anthropic-auth/pull/109?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

<details><summary><h3>Greptile Summary</h3></summary>

This PR adds a `scoped` third dimension to the per-account killswitch, enabling model-specific weekly quota enforcement (e.g., Fable carve-outs) without blocking the account for other models. The change threads `modelId` through the main killswitch gate and both fallback filters, and introduces scope-aware block-message attribution and retry-after hints.

- **`killswitchPassesPolicy`** gains an optional `modelId`; when present and a scoped window matches at/below the `scoped` threshold, the request is blocked independently of 5h/7d — using `<=` (inclusive) so the default `0` fires at true exhaustion.
- **`resolveScopedDrivenBlock`** distinguishes a scoped-driven block from a 5h/7d-driven one by re-running policy without `modelId`; if 5h/7d already killed the account, the block is classified as account-level regardless of the scoped window's state (addressing the previous FINDING 2).
- **`killswitchRetryAfterSeconds`** now takes a `scopedModelId` and, when provided, collects only the matched scoped window's weekly `resetsAt` — ignoring the 5h reset that would otherwise produce a misleadingly short retry hint (addressing the previous FINDING 1).
</details>

<h3>Confidence Score: 5/5</h3>

Safe to merge — the two previously flagged issues (retry-after dominated by 5h/7d resets; misclassification when both 5h/7d and scoped are simultaneously exhausted) are both resolved in this revision, with regression-lock tests co-located alongside each fix.

The scoped check is correctly ordered before the unknown-window fail-open branch, the resolveScopedDrivenBlock priority logic (5h/7d wins over scoped when both are exhausted) matches the stated invariant and is directly verified, and the retry-after split (scoped-only vs 5h/7d) prevents the retry-storm scenario. The change is backward-compatible: existing two-value configs read scoped as 0 on normalization, and callers that omit modelId see identical behavior to before.

No files require special attention.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/core/src/accounts.ts | Adds scoped to KillswitchThresholds, DEFAULT_KILLSWITCH_THRESHOLDS, and normalizeKillswitchThresholds; threads optional modelId into killswitchPassesPolicy with correct ordering (before the unknown-window fail-open branch); splits killswitchRetryAfterSeconds into scoped-only vs 5h/7d paths — all sound. |
| packages/core/src/killswitch.ts | Extends the command parser regex to accept an optional third number, updates the set-entry type, and adds a Scoped column to the status table via normalizeKillswitchThresholds — backward-compatible and clean. |
| packages/opencode/src/index.ts | Introduces resolveScopedDrivenBlock and formatKillswitchBlockMessage; threads modelId into the main gate, getRoutableFallbackAccounts, and the second fallback filter; correctly conditions retry-after and block message on isScopedDriven — logic is consistent with the policy functions. |
| packages/core/src/tests/killswitch.test.ts | New test file with 34 cases covering defaults, normalization, per-model isolation, boundary conditions (<=), MUST-1 fail-open ordering, FINDING 1 retry-after branching, and config round-trips — comprehensive. |
| packages/opencode/src/tests/killswitch.test.ts | Adds formatKillswitchBlockMessage and resolveScopedDrivenBlock tests covering MUST-2 (5h/7d priority), FINDING 2 regression lock, and message attribution; thoroughly exercises the correct block-attribution logic. |

</details>

<details><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Request with modelId] --> B{killswitch enabled?}
    B -- No --> Z[Pass through to main]
    B -- Yes --> C[killswitchPassesPolicy
mainQuota, storage, undefined, modelId]
    C --> D{5h/7d window
below threshold?}
    D -- Yes return false --> E[Account-level kill]
    D -- No check scoped --> F{modelId provided
and scoped window
matches model?}
    F -- No scoped window --> G{sawUnknownWindow?}
    F -- Window at/below
scoped threshold --> E
    F -- Window above
scoped threshold --> G
    G -- failClosed=true --> E
    G -- failClosed=false --> Z
    E --> H[getRoutableFallbackAccounts
filtered by modelId scoped check]
    H --> I{survivingFallbacks > 0?}
    I -- Yes --> J[Route to fallback]
    I -- No --> K[resolveScopedDrivenBlock
check 5h/7d WITHOUT modelId]
    K --> L{5h/7d passes
without modelId?}
    L -- No account-level --> M[Generic 429 message
retryAfter from 5h/7d resets]
    L -- Yes + scoped window
at/below threshold --> N[Scoped-driven 429
model-named message
retryAfter from weekly resetsAt]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[Request with modelId] --> B{killswitch enabled?}
    B -- No --> Z[Pass through to main]
    B -- Yes --> C[killswitchPassesPolicy
mainQuota, storage, undefined, modelId]
    C --> D{5h/7d window
below threshold?}
    D -- Yes return false --> E[Account-level kill]
    D -- No check scoped --> F{modelId provided
and scoped window
matches model?}
    F -- No scoped window --> G{sawUnknownWindow?}
    F -- Window at/below
scoped threshold --> E
    F -- Window above
scoped threshold --> G
    G -- failClosed=true --> E
    G -- failClosed=false --> Z
    E --> H[getRoutableFallbackAccounts
filtered by modelId scoped check]
    H --> I{survivingFallbacks > 0?}
    I -- Yes --> J[Route to fallback]
    I -- No --> K[resolveScopedDrivenBlock
check 5h/7d WITHOUT modelId]
    K --> L{5h/7d passes
without modelId?}
    L -- No account-level --> M[Generic 429 message
retryAfter from 5h/7d resets]
    L -- Yes + scoped window
at/below threshold --> N[Scoped-driven 429
model-named message
retryAfter from weekly resetsAt]
```

</a>
</details>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `packages/opencode/src/index.ts`, line 886-911 ([link](https://github.com/cortexkit/anthropic-auth/blob/5209727bfbcf6c97574be472d19cb2a6ac20e924/packages/opencode/src/index.ts#L886-L911)) 

   <a href="#">P1</a> **`resolveScopedDrivenBlock` misclassifies when 5h/7d and scoped are both exhausted**

   `resolveScopedDrivenBlock` checks only whether the scoped window is at/below threshold — it does not verify that 5h/7d windows are themselves healthy. When 5h/7d exhaustion (e.g. `remainingPercent: 4` vs a threshold of `5`) caused `killswitchPassesPolicy` to return `false`, a simultaneously exhausted scoped window still produces `{ isScopedDriven: true }`. The caller then emits `"<Model> weekly limit reached, no routable accounts"`, misleading operators into thinking Sonnet requests on the same account would succeed — when in reality the account is blocked for all models due to the 5h/7d breach.

   The fix is to confirm that no 5h/7d window is independently below its threshold before declaring the block scoped-driven: if any finite 5h/7d window is `< thresholds[key]`, return `{ isScopedDriven: false }` immediately so the generic account-level message is used instead.
</details>

<sub>Reviews (3): Last reviewed commit: ["test(killswitch): make MUST-2 5h/7d prem..."](https://github.com/cortexkit/anthropic-auth/commit/660e47dbc49699c957ed82ca36d3b8be8a9d380e) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=41893911)</sub>

### CODE REVIEW

File: packages/core/src/accounts.ts [hunk 1]
```diff
@@ -1744,6 +1781,15 @@ export function killswitchRetryAfterSeconds(
...
+    if (scopedModelId) {
+      const scopedWindow = getScopedQuotaWindowForModel(quota, scopedModelId)
+      const resetStr = scopedWindow?.resetsAt
+      if (!resetStr) continue
+      const resetTime = Date.parse(resetStr)
+      if (Number.isFinite(resetTime) && resetTime > now) {
+        resetTimes.push(resetTime)
+      }
+    }
   }
```
[the remark this answers is not in the document]
Comment (iceteaSA) [reply packages/core/src/accounts.ts:1796]: Fixed in 70346d0. `killswitchRetryAfterSeconds` now uses an `if (scopedModelId) { …scoped only… } else { …5h/7d… }` split — for a scoped-driven block only the matched scoped window's weekly `resetsAt` is collected, so the retry hint reflects the weekly reset instead of the sooner 5h reset. The no-scoped path is byte-identical. Red-first test co-locates a +2h 5h reset and a +2-day Fable reset in the same quota (the real retry-storm shape) and asserts ~2 days; it failed (~2h) before the fix.

File: packages/core/src/accounts.ts [hunk 2]
```diff
@@ -1744,6 +1781,15 @@ export function killswitchRetryAfterSeconds(
         resetTimes.push(resetTime)
       }
     }
+    if (scopedModelId) {
```
[the remark this answers is not in the document]
Comment (iceteaSA) [reply packages/core/src/accounts.ts:1778]: Fixed in 70346d0 (same fix as the greptile P1 on this function). When `scopedModelId` is provided, only the matched scoped window's `resetsAt` is considered; the 5h/7d collection is skipped entirely for that path, so `Math.min` can no longer return a sooner 5h reset for a weekly scoped block.

File: packages/opencode/src/index.ts [hunk 3]
```diff
@@ -153,6 +155,60 @@ const MIN_MAIN_REFRESH_BEFORE_EXPIRY_MINUTES = 240
...
+
+/**
+ * Decide whether a killswitch block is scoped-driven (a per-model weekly
+ * block) versus a whole-account 5h/7d-driven block. A block is scoped-driven
+ * only when the request's model matches a scoped window AND that window is
+ * actually at or below the per-account scoped threshold. A Fable request
+ * with a healthy Fable window is therefore correctly classified as
+ * account-level (the 5h/7d breach killed the account, not the Fable quota).
+ */
+export function resolveScopedDrivenBlock(input: {
```
[the remark this answers is not in the document]
Comment (iceteaSA) [reply packages/opencode/src/index.ts:191]: Fixed in 70346d0. `resolveScopedDrivenBlock` now guards on the core policy before the scoped check: `if (!killswitchPassesPolicy(mainQuota, storage)) return { isScopedDriven: false }`. Called without a modelId, `killswitchPassesPolicy` evaluates only 5h/7d — so if the account is 5h/7d-killed, the block is attributed account-level regardless of the scoped window. Reusing the core function (rather than re-deriving the 5h/7d logic in opencode) also addresses the cross-package-drift concern. Red-first test: both 5h/7d and Fable exhausted → now `isScopedDriven: false` (was `true`); the healthy-5h/7d + exhausted-Fable case still returns `true`.

File: packages/opencode/src/tests/killswitch.test.ts [hunk 4]
```diff
@@ -535,3 +542,343 @@ describe('getQuotaRefreshEveryNRequests', () => {
...
+    // Main quota is exhausted on 5h/7d (killswitch will fire), but the
+    // Fable window is at 100% remaining — well above the scoped threshold.
+    // The current (buggy) code names Fable; the fix must produce a generic
+    // account-level message because the block is 5h/7d-driven, not
+    // scoped-driven.
+    const storage = killswitchStorage(0)
+    const mainQuota: OAuthQuotaSnapshot = {
+      five_hour: {
+        usedPercent: 95,
+        remainingPercent: 5,
```
[the remark this answers is not in the document]
Comment (iceteaSA) [reply packages/opencode/src/tests/killswitch.test.ts]: Valid — fixed in 660e47d. The test set `five_hour.remainingPercent: 5` / `seven_day: 10` against thresholds `5`/`10`, and since 5h/7d use strict `<`, `5 < 5` and `10 < 10` are both false → `killswitchPassesPolicy` returned true (not blocked), so the account-level priority guard never executed. Lowered to `4`/`9` so the killswitch genuinely fires and the test traverses the guard branch.

Note the both-exhausted case (5h/7d AND Fable both below threshold — the path where the guard is the *only* thing producing `isScopedDriven: false`) is already covered red-first by the existing `BOTH 5h/7d AND Fable exhausted → NOT scoped-driven (FINDING 2)` test (thresholds 10/10 vs remaining 4/4), so this change is just correcting the healthy-Fable variant's premise rather than adding a duplicate.
